Improve AMP sanitization for title tags and meta descriptions

swissspidy opened the pull request:
## Context

<!-- What do we want to achieve with this PR? Why did we write this code? -->

We had issues with missing title tags / meta descriptions or duplicate ones, depending on the theme and SEO plugin being used.

I also contributed a [PR to Yoast SEO](https://github.com/Yoast/wordpress-seo/pull/19214) to improve their logic, as it turns out they did some odd changes there recently.

## Summary

<!-- A brief description of what this PR does. -->

Prevent missing title tags / meta descriptions or duplicate ones at the AMP sanitization level.

Also added another fix to ensure links always have a proper protocol.

## Relevant Technical Choices

<!-- Please describe your changes. -->

## To-do

<!-- A list of things that need to be addressed in this PR or follow-up changes. -->

## User-facing changes

<!--
Please describe your changes.
Include before/after screenshots or a short video.
-->

No visual changes

## Testing Instructions

<!--
How can the changes in this PR be verified?
Please provide step-by-step instructions how to reproduce the issue, if applicable.
Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->

<!-- ignore-task-list-start -->
- [ ] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

This PR can be tested by following these steps:

1. View a published story on the frontend
2. Verify that there's a `<title>` tag and a meta description present

Test with:

* Twenty Twelve theme
* Twenty Twenty-Three theme
* OceanWP theme
* Yoast SEO enabled/disabled

## Reviews

### Does this PR have a security-related impact?

<!-- Examples: new APIs, changes to KSES, etc.  -->

No

### Does this PR change what data or activity we track or use?

<!-- Examples: changes to telemetry, new third-party APIs -->

No

### Does this PR have a legal-related impact?

<!-- Examples: new images with unknown sources, new production dependencies with incompatible licenses -->

No

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.

NOTE: One reference per line!

Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #12655
Fixes #12709

swissspidy committed:
Sanitize links with missing protocol

swissspidy committed:
Always print metadata & document title

swissspidy committed:
Sanitize `<title>` tags and meta descriptions

Prevent duplicates, add missing ones

swissspidy committed:
Lint & test fixes

spacedmonkey commented on includes/AMP/Sanitization.php:
```diff
@@ -442,6 +442,8 @@ protected function get_sanitizers(): array {
 				'poster_images'  => array_filter( $poster_images ),
 				'video_cache'    => $video_cache_enabled,
+				'title_tag'      => wp_get_document_title(),
+				'description'    => wp_strip_all_tags( get_the_excerpt() ),
```
What about excerpt length and `wp_trim_excerpt`? Adding `[&hellip;]` at the end here might not be the best thing.

spacedmonkey replied:
Why not use `$story->get_excerpt()`?

swissspidy replied:
This is copied from the Discovery class. Has the advantage that it has a maximum length and is filterable.

spacedmonkey replied:
Is clipping the text a good thing in this place?

swissspidy replied:
It's the existing behavior of the plugin.

I don't see an issue with clipping. While there's no limit on how long a meta description can be, the snippet is typically truncated in Google Search results as needed anyway.

spacedmonkey replied:
Okay, cool.Just checking.

spacedmonkey commented on includes/AMP/Traits/Sanitization_Utils.php:
```diff
@@ -113,6 +115,17 @@ private function transform_a_tags( $document ): void {
+				0 !== stripos( $url, 'http://' ) &&
+				0 !== stripos( $url, 'https://' ) &&
+				0 !== stripos( $url, 'tel:' ) &&
+				0 !== stripos( $url, 'mailto:' )
```
Could we use `wp_allowed_protocols()` here?

spacedmonkey replied:
Maybe look at `wp_kses_bad_protocol_once`.

swissspidy replied:
This matches the list of allowed protocols in the editor, as per the inline comment.

spacedmonkey replied:
Okay, how about an array you loop around then?

spacedmonkey replied:
Again we can use `str_starts_with` here.

swissspidy replied:
You mean something like this?

```php
$has_protocol = false;
foreach ( [ 'http://', 'https://', 'tel:', 'mailto:' ] as $protocol ) {
	if ( 0 === stripos( $url, $protocol ) ) {
		$has_protocol = true;
		break;
	}
}

if ( ! $has_protocol ) {
	$link->setAttribute( 'href', 'https://' . $url );
}
```

That is much longer and less readable

spacedmonkey replied:
I find the more readable. But we have had this discussion before..

spacedmonkey commented on includes/AMP/Traits/Sanitization_Utils.php:
```diff
@@ -613,4 +626,83 @@ private function remove_page_template_placeholder_images( $document ): void {
+		 */
+		$meta_descriptions = $document->xpath->query( './/meta[@name="description"]' );
+
+		if ( $meta_descriptions->length > 1 ) {
+			foreach ( $meta_descriptions as $index => $meta_description ) {
+				if ( 0 === $index ) {
+					continue;
+				}
+				$document->head->removeChild( $meta_description );
+			}
+		}
```
These lines are repeated from above. https://github.com/GoogleForCreators/web-stories-wp/pull/12706/files#diff-56d46ddccbd78e7364b5120cf250aff849e7af145759193537cfcb8c3c27fdfeR651-R658. Move this logic to a method.

swissspidy replied:
A bit overkill IMO to add an extra method for this

spacedmonkey replied:
It makes it more testable....

spacedmonkey replied:
And reduces complexity of method.

spacedmonkey commented on includes/Integrations/AMP.php:
```diff
@@ -234,6 +234,8 @@ public function add_amp_content_sanitizers( $sanitizers ) {
 			'poster_images'  => array_filter( $poster_images ),
 			'video_cache'    => $video_cache_enabled,
+			'title_tag'      => wp_get_document_title(),
+			'description'    => wp_strip_all_tags( get_the_excerpt() ),
```
See https://github.com/GoogleForCreators/web-stories-wp/pull/12706/files#r1026440425

spacedmonkey commented on includes/AMP/Traits/Sanitization_Utils.php:
```diff
@@ -88,7 +88,9 @@ private function transform_a_tags( $document ): void {
-			$is_link_to_same_origin = 0 === strpos( $link->getAttribute( 'href' ), home_url() );
+			$url = $link->getAttribute( 'href' );
+
+			$is_link_to_same_origin = 0 === strpos( $url, home_url() );
```
How about using `str_starts_with`? There is a polyfill in core for this.

swissspidy replied:
Oh neat, I didn't know these were added in WordPress 5.9. I thought they were added later.

Makes total sense in that case!

spacedmonkey commented on includes/Discovery.php:
```diff
@@ -103,26 +103,25 @@ public function register(): void {
 		 * @param bool $enable_open_graph Whether to print the document title. Defaults to true for block themes, false otherwise.
 		 */
-		$enable_metadata = apply_filters( 'web_stories_enable_document_title', $enable_metadata );
+		$enable_metadata = apply_filters( 'web_stories_enable_document_title', true );
```
Update the comment above?

swissspidy committed:
Update docblock

swissspidy committed:
Use `str_contains` and `str_starts_with` where applicable

WordPress 5.9+ adds polyfills for these PHP8+ functions, see https://make.wordpress.org/core/2022/01/10/wordpress-5-9-and-php-8-0-8-1/

swissspidy committed:
Remove now obsolete tests

spacedmonkey reviewed (approved):
I think we are looking good here.

spacedmonkey committed:
Merge branch 'main' into fix/12655-title-meta-output
